types/proposal.go

```diff
@@ -89,11 +90,25 @@ func (p *Proposal) ValidateBasic() error {
 //
 // For more information on the meaning of 'timely', see the proposer-based timestamp specification:
 // https://github.com/tendermint/spec/tree/master/spec/consensus/proposer-based-timestamp
-func (p *Proposal) IsTimely(recvTime time.Time, sp SynchronyParams) bool {
+func (p *Proposal) IsTimely(recvTime time.Time, sp SynchronyParams, round int32) bool {
+	// The message delay values are scaled as rounds progress.
+	// Every 10 rounds, the message delay is doubled to allow consensus to
+	// proceed in the case that the chosen value was too small for the given network conditions.
+	// For more information and discussion on this mechanism, see the relevant github issue:
+	// https://github.com/tendermint/spec/issues/371
+	roundModifier := time.Duration(math.Exp2(float64(round / 10)))
+	msgDelay := sp.MessageDelay * roundModifier
```

Comment:
Optionally, you could avoid the overflow check and the floating point arithmetic, if you precompute a maximum shift:

```go
// import "math/bits"
maxShift := bits.LeadingZeros64(uint64(sp.MessageDelay)) - 1
```

That is how far left you can shift sp.MessageDelay left, i.e., how many powers of two you can multiply it by, before you would overflow. Given that:

```go
	nshift := (round / 10)
	if nshift > maxShift {
		nshift = maxShift
	}
	msgDelay := sp.MessageDelay * time.Duration(1<<nshift)
```

This will never overflow, and requires only integer arithmetic. Even if you recompute `maxShift` each time it's probably more efficient (`LeadingZeros64` does the usual divide-and-conquer thing to minimize shifts).

YMMV, what you have here is totally fine too.

Reply:
This is interesting, I've never used `bits.LeadingZeros` so this is a good excuse to do so.

Reply:
This isn't really a context where performance matters all that much, the main reason it caught my attention was the combination of floating-point and wraparound checking 🙂